File: ios/OCR/Modules/OCRModule.swift
import Foundation
import UIKit
import React
import Vision

/**
 * OCR Module - Secure implementation with token-based data exchange
 *
 * Fixes: SEC-001 - PII leakage through React Native bridge
 *
 * Changes:
 * - scan() now returns only { sessionToken, confidence, status }
 * - Actual PII data stored securely in native with 5-minute TTL
 * - New method getResult(token) retrieves the actual data
 */
@objc(OCRModule)
class OCRModule: RCTEventEmitter {
    
    private let secureStorage = SecureStorage.shared
    private let minConfidenceThreshold: Float = 0.85
    
    override static func moduleName() -> String! {
        return "OCRModule"
    }
    
    override func supportedEvents() -> [String]! {
        return ["OCR_SUCCESS", "OCR_ERROR", "OCR_PROGRESS"]
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    // MARK: - Public Methods
    
    /**
     * Scan document and extract text
     * Returns only sessionToken + metadata (no PII data)
     */
    @objc func scan(
        _ imageUri: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            do {
                // Load image
                guard let image = self.loadImage(from: imageUri) else {
                    reject("OCR_INVALID_IMAGE", "Failed to load image", nil)
                    return
                }
                
                // Perform OCR
                let ocrResult = try self.performOCR(image: image)
                
                // Validate confidence threshold
                if ocrResult.confidence < self.minConfidenceThreshold {
                    let errorInfo: [String: Any] = [
                        "code": "OCR_001",
                        "message": "Low confidence score",
                        "confidence": ocrResult.confidence,
                        "retryable": true
                    ]
                    reject(
                        "OCR_001",
                        "Low confidence score: \(ocrResult.confidence)",
                        NSError(domain: "OCRModule", code: 1, userInfo: errorInfo)
                    )
                    return
                }
                
                // Generate session token
                let sessionToken = self.secureStorage.generateToken()
                
                // Store PII data securely (5 minute TTL)
                try self.secureStorage.save(sessionToken, data: ocrResult, ttl: 300)
                
                // Return only token + metadata (NO PII)
                let response: [String: Any] = [
                    "status": "SUCCESS",
                    "sessionToken": sessionToken,
                    "confidence": ocrResult.confidence,
                    "metadata": [
                        "processingTime": ocrResult.processingTime,
                        "imageQuality": ocrResult.imageQuality,
                        "hasGlare": ocrResult.hasGlare,
                        "isBlurry": ocrResult.isBlurry,
                        "lightingCondition": ocrResult.lightingCondition
                    ]
                ]
                
                resolve(response)
                
                // Emit event for analytics (no PII)
                self.sendEvent(
                    withName: "OCR_SUCCESS",
                    body: [
                        "sessionToken": sessionToken,
                        "confidence": ocrResult.confidence
                    ]
                )
                
            } catch {
                reject("OCR_999", "Unexpected error: \(error.localizedDescription)", error)
            }
        }
    }
    
    /**
     * Retrieve OCR result by session token
     * This is the ONLY method that returns PII data
     */
    @objc func getResult(
        _ sessionToken: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            do {
                // Validate token format
                guard !sessionToken.isEmpty else {
                    reject("OCR_INVALID_TOKEN", "Session token is required", nil)
                    return
                }
                
                // Retrieve data from secure storage
                guard let ocrResult = try self.secureStorage.get(sessionToken, as: OCRResult.self) else {
                    let errorInfo: [String: Any] = [
                        "code": "OCR_TOKEN_EXPIRED",
                        "message": "Session token expired or invalid (5 minute TTL)",
                        "confidence": 0.0,
                        "retryable": false
                    ]
                    reject(
                        "OCR_TOKEN_EXPIRED",
                        "Session token expired or invalid",
                        NSError(domain: "OCRModule", code: 2, userInfo: errorInfo)
                    )
                    return
                }
                
                // Delete token after retrieval (one-time use)
                try self.secureStorage.delete(sessionToken)
                
                // Return full data including PII
                let response: [String: Any] = [
                    "status": ocrResult.status,
                    "confidence": ocrResult.confidence,
                    "fields": [
                        "tcNo": [
                            "value": ocrResult.fields.tcNo.value,
                            "confidence": ocrResult.fields.tcNo.confidence
                        ],
                        "name": [
                            "value": ocrResult.fields.name.value,
                            "confidence": ocrResult.fields.name.confidence
                        ],
                        "surname": [
                            "value": ocrResult.fields.surname.value,
                            "confidence": ocrResult.fields.surname.confidence
                        ],
                        "birthDate": [
                            "value": ocrResult.fields.birthDate.value,
                            "confidence": ocrResult.fields.birthDate.confidence
                        ],
                        "documentNo": [
                            "value": ocrResult.fields.documentNo.value,
                            "confidence": ocrResult.fields.documentNo.confidence
                        ]
                    ],
                    "metadata": [
                        "processingTime": ocrResult.processingTime,
                        "imageQuality": ocrResult.imageQuality,
                        "hasGlare": ocrResult.hasGlare,
                        "isBlurry": ocrResult.isBlurry,
                        "lightingCondition": ocrResult.lightingCondition
                    ]
                ]
                
                resolve(response)
                
            } catch {
                reject("OCR_RETRIEVAL_ERROR", "Failed to retrieve result: \(error.localizedDescription)", error)
            }
        }
    }
    
    /**
     * Validate if a session token is still valid
     */
    @objc func isTokenValid(
        _ sessionToken: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        let isValid = secureStorage.isValid(sessionToken)
        resolve(isValid)
    }
    
    /**
     * Manually cleanup expired tokens
     */
    @objc func cleanupExpiredTokens(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        secureStorage.cleanupExpired()
        resolve(true)
    }
    
    // MARK: - Private Helper Methods
    
    private func loadImage(from uri: String) -> UIImage? {
        // Handle file:// URIs
        let cleanUri = uri.replacingOccurrences(of: "file://", with: "")
        
        if let image = UIImage(contentsOfFile: cleanUri) {
            return image
        }
        
        // Try as regular path
        return UIImage(contentsOfFile: uri)
    }
    
    private func performOCR(image: UIImage) throws -> OCRResult {
        let startTime = Date()
        
        // Initialize frame processor
        let frameProcessor = FrameProcessor()
        
        // Step 1: Analyze image quality
        let quality = frameProcessor.analyzeQuality(image)
        
        // Check if image quality is acceptable
        guard quality.overallScore > 0.5 else {
            throw OCRError.poorImageQuality
        }
        
        // Step 2: Preprocess image
        guard let preprocessedImage = frameProcessor.preprocessForOCR(image) else {
            throw OCRError.preprocessingFailed
        }
        
        // Step 3: Perform OCR using Vision framework
        guard let cgImage = preprocessedImage.cgImage else {
            throw OCRError.invalidImage
        }
        
        var extractedText = ""
        var recognitionConfidence: Float = 0.0
        
        let semaphore = DispatchSemaphore(value: 0)
        var ocrError: Error?
        
        // Create Vision request
        let request = VNRecognizeTextRequest { request, error in
            if let error = error {
                ocrError = error
                semaphore.signal()
                return
            }
            
            guard let observations = request.results as? [VNRecognizedTextObservation] else {
                semaphore.signal()
                return
            }
            
            // Extract text from observations
            var allText: [String] = []
            var confidenceSum: Float = 0.0
            
            for observation in observations {
                guard let topCandidate = observation.topCandidates(1).first else { continue }
                allText.append(topCandidate.string)
                confidenceSum += topCandidate.confidence
            }
            
            extractedText = allText.joined(separator: "\n")
            recognitionConfidence = observations.isEmpty ? 0.0 : confidenceSum / Float(observations.count)
            
            semaphore.signal()
        }
        
        // Configure request
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true
        request.recognitionLanguages = ["tr-TR", "en-US"] // Turkish and English
        
        // Perform request
        let requestHandler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        
        do {
            try requestHandler.perform([request])
        } catch {
            throw OCRError.visionRequestFailed
        }
        
        // Wait for completion
        semaphore.wait()
        
        if let error = ocrError {
            throw error
        }
        
        // Step 4: Parse extracted fields from text
        let fields = parseFields(from: extractedText, confidence: recognitionConfidence)
        
        let processingTime = Int(Date().timeIntervalSince(startTime) * 1000)
        
        // Calculate overall confidence
        let overallConfidence = (recognitionConfidence + Float(quality.overallScore)) / 2.0
        
        // Determine lighting condition
        let lightingCondition: String
        if quality.brightness < 0.3 {
            lightingCondition = "DARK"
        } else if quality.brightness > 0.8 {
            lightingCondition = "BRIGHT"
        } else {
            lightingCondition = "GOOD"
        }
        
        return OCRResult(
            status: "SUCCESS",
            confidence: overallConfidence,
            fields: fields,
            processingTime: processingTime,
            imageQuality: Float(quality.overallScore),
            hasGlare: quality.hasGlare,
            isBlurry: quality.isBlurry,
            lightingCondition: lightingCondition
        )
    }
    
    /**
     * Parse OCR fields from extracted text
     * This uses regex patterns to identify Turkish ID card fields
     */
    private func parseFields(from text: String, confidence: Float) -> OCRFields {
        var tcNo = ""
        var name = ""
        var surname = ""
        var birthDate = ""
        var documentNo = ""
        
        // Parse TC No (11 digits)
        if let tcMatch = text.range(of: "\\b\\d{11}\\b", options: .regularExpression) {
            tcNo = String(text[tcMatch])
        }
        
        // Parse Name (after "ADI" or "NAME" keyword)
        if let nameMatch = text.range(of: "(?:ADI|NAME)[:\\s]+([A-ZÇĞİÖŞÜ]+)", options: .regularExpression) {
            name = String(text[nameMatch]).replacingOccurrences(of: "ADI", with: "")
                .replacingOccurrences(of: "NAME", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
        
        // Parse Surname (after "SOYADI" or "SURNAME" keyword)
        if let surnameMatch = text.range(of: "(?:SOYADI|SURNAME)[:\\s]+([A-ZÇĞİÖŞÜ]+)", options: .regularExpression) {
            surname = String(text[surnameMatch]).replacingOccurrences(of: "SOYADI", with: "")
                .replacingOccurrences(of: "SURNAME", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
        
        // Parse Birth Date (DD.MM.YYYY format)
        if let dateMatch = text.range(of: "\\b\\d{2}\\.\\d{2}\\.\\d{4}\\b", options: .regularExpression) {
            birthDate = String(text[dateMatch])
        }
        
        // Parse Document No (letter + digits)
        if let docMatch = text.range(of: "\\b[A-Z]\\d{8}\\b", options: .regularExpression) {
            documentNo = String(text[docMatch])
        }
        
        return OCRFields(
            tcNo: FieldData(value: tcNo, confidence: confidence),
            name: FieldData(value: name, confidence: confidence * 0.95),
            surname: FieldData(value: surname, confidence: confidence * 0.95),
            birthDate: FieldData(value: birthDate, confidence: confidence * 0.93),
            documentNo: FieldData(value: documentNo, confidence: confidence * 0.94)
        )
    }
}

// MARK: - Data Models

struct OCRResult: Codable {
    let status: String
    let confidence: Float
    let fields: OCRFields
    let processingTime: Int
    let imageQuality: Float
    let hasGlare: Bool
    let isBlurry: Bool
    let lightingCondition: String
}

struct OCRFields: Codable {
    let tcNo: FieldData
    let name: FieldData
    let surname: FieldData
    let birthDate: FieldData
    let documentNo: FieldData
}

struct FieldData: Codable {
    let value: String
    let confidence: Float
}

// MARK: - Error Types

enum OCRError: Error {
    case invalidImage
    case poorImageQuality
    case preprocessingFailed
    case visionRequestFailed
    case noTextDetected
    
    var localizedDescription: String {
        switch self {
        case .invalidImage:
            return "Invalid image format"
        case .poorImageQuality:
            return "Image quality too poor for OCR"
        case .preprocessingFailed:
            return "Failed to preprocess image"
        case .visionRequestFailed:
            return "Vision framework request failed"
        case .noTextDetected:
            return "No text detected in image"
        }
    }
}

File: android/app/src/main/java/com/ocr/modules/OCRModule.java
package com.ocr.modules;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.facebook.react.bridge.*;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.ocr.security.SecureStorage;
import com.ocr.utils.FrameProcessor;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * OCR Module - Java implementation with secure token-based data exchange
 * 
 * Fixes:
 * - SEC-001: PII leakage through React Native bridge
 * - BRIDGE-001: Memory leak in frame processing
 * 
 * Features:
 * - scan() returns only sessionToken + metadata (NO PII)
 * - getResult() retrieves PII securely (one-time use)
 * - Memory metrics tracking
 * - ExecutorService for async operations
 * 
 * Kotlin→Java conversion:
 * - Coroutines → ExecutorService
 * - Promise callbacks on UI thread
 * - Manual resource cleanup
 */
@ReactModule(name = "OCRModule")
public class OCRModule extends ReactContextBaseJavaModule {
    
    private static final String TAG = "OCRModule";
    private static final double MIN_CONFIDENCE_THRESHOLD = 0.85;
    
    private final SecureStorage secureStorage;
    private final FrameProcessor frameProcessor;
    private final ExecutorService executorService;
    private final Handler mainHandler;
    
    public OCRModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.secureStorage = SecureStorage.getInstance(reactContext);
        this.frameProcessor = new FrameProcessor();
        this.executorService = Executors.newFixedThreadPool(2);
        this.mainHandler = new Handler(Looper.getMainLooper());
    }
    
    @Override
    public String getName() {
        return "OCRModule";
    }
    
    /**
     * Scan document and extract text
     * Returns only sessionToken + metadata + memory metrics (NO PII)
     * 
     * @param imageUri Path to image file
     * @param promise React Native promise
     */
    @ReactMethod
    public void scan(String imageUri, Promise promise) {
        executorService.submit(() -> {
            try {
                Log.d(TAG, "Starting OCR scan for: " + imageUri);
                
                // Load image
                Bitmap bitmap = BitmapFactory.decodeFile(imageUri);
                if (bitmap == null) {
                    rejectOnUIThread(promise, "OCR_INVALID_IMAGE", "Failed to load image", 
                            createErrorMap("OCR_INVALID_IMAGE", "Invalid image file", 0.0, true));
                    return;
                }
                
                // Process with automatic cleanup and memory tracking
                Future<FrameProcessor.ProcessingResult<OCRResult>> future = 
                        frameProcessor.processBitmap(bitmap, this::performOCR);
                
                FrameProcessor.ProcessingResult<OCRResult> processingResult = future.get();
                OCRResult ocrResult = processingResult.data;
                FrameProcessor.ProcessingMetrics metrics = processingResult.metrics;
                
                Log.d(TAG, String.format("OCR completed: confidence=%.2f, time=%dms, memory=%.2fMB→%.2fMB",
                        ocrResult.confidence, metrics.processingTimeMs, 
                        metrics.memoryUsedBeforeMB, metrics.memoryUsedAfterMB));
                
                // Check confidence threshold
                if (ocrResult.confidence < MIN_CONFIDENCE_THRESHOLD) {
                    Log.w(TAG, "Low confidence: " + ocrResult.confidence);
                    rejectOnUIThread(promise, "OCR_001", 
                            "Low confidence score: " + ocrResult.confidence,
                            createErrorMap("OCR_001", "Low confidence score", ocrResult.confidence, true));
                    return;
                }
                
                // Generate session token
                String sessionToken = secureStorage.generateToken();
                
                // Store PII data securely (5 minute TTL)
                secureStorage.save(sessionToken, ocrResult, 300_000L);
                
                // Get current memory stats
                FrameProcessor.MemoryStats memoryStats = frameProcessor.getMemoryStats();
                
                // Return only token + metadata + memory metrics (NO PII)
                WritableMap response = Arguments.createMap();
                response.putString("status", "SUCCESS");
                response.putString("sessionToken", sessionToken);
                response.putDouble("confidence", ocrResult.confidence);
                
                WritableMap metadata = Arguments.createMap();
                metadata.putString("documentType", "ID_CARD");
                metadata.putInt("fieldCount", 5);
                
                // Memory metrics
                WritableMap performance = Arguments.createMap();
                performance.putDouble("processingTimeMs", (double) metrics.processingTimeMs);
                performance.putDouble("memoryUsedBeforeMB", metrics.memoryUsedBeforeMB);
                performance.putDouble("memoryUsedAfterMB", metrics.memoryUsedAfterMB);
                performance.putDouble("memoryFreedMB", metrics.memoryFreedMB);
                performance.putDouble("bitmapSizeMB", metrics.bitmapSizeBytes / (1024.0 * 1024.0));
                performance.putDouble("currentMemoryUsageMB", memoryStats.usedMemoryMB);
                performance.putDouble("memoryUsagePercent", memoryStats.usagePercent);
                
                metadata.putMap("performance", performance);
                response.putMap("metadata", metadata);
                
                Log.d(TAG, "OCR scan successful: token=" + sessionToken);
                resolveOnUIThread(promise, response);
                
                // Emit success event (no PII)
                WritableMap eventData = Arguments.createMap();
                eventData.putString("sessionToken", sessionToken);
                eventData.putDouble("confidence", ocrResult.confidence);
                eventData.putDouble("processingTimeMs", (double) metrics.processingTimeMs);
                sendEvent("OCR_SUCCESS", eventData);
                
            } catch (Exception e) {
                Log.e(TAG, "OCR scan error", e);
                rejectOnUIThread(promise, "OCR_999", "Unexpected error: " + e.getMessage(), null);
            }
        });
    }
    
    /**
     * Retrieve OCR result by session token
     * This is the ONLY method that returns PII data
     * 
     * @param sessionToken Session token from scan()
     * @param promise React Native promise
     */
    @ReactMethod
    public void getResult(String sessionToken, Promise promise) {
        executorService.submit(() -> {
            try {
                if (sessionToken == null || sessionToken.isEmpty()) {
                    rejectOnUIThread(promise, "OCR_INVALID_TOKEN", "Session token is required", null);
                    return;
                }
                
                // Retrieve from secure storage
                OCRResult ocrResult = secureStorage.get(sessionToken, OCRResult.class);
                
                if (ocrResult == null) {
                    rejectOnUIThread(promise, "OCR_TOKEN_EXPIRED",
                            "Session token expired or invalid",
                            createErrorMap("OCR_TOKEN_EXPIRED", "Token expired (5 minute TTL)", 0.0, false));
                    return;
                }
                
                // Delete token after retrieval (one-time use)
                secureStorage.delete(sessionToken);
                
                // Return full data including PII
                WritableMap response = Arguments.createMap();
                response.putString("status", ocrResult.status);
                response.putDouble("confidence", ocrResult.confidence);
                
                WritableMap fields = Arguments.createMap();
                fields.putMap("tcNo", createFieldDataMap(ocrResult.fields.tcNo));
                fields.putMap("name", createFieldDataMap(ocrResult.fields.name));
                fields.putMap("surname", createFieldDataMap(ocrResult.fields.surname));
                fields.putMap("birthDate", createFieldDataMap(ocrResult.fields.birthDate));
                fields.putMap("serialNumber", createFieldDataMap(ocrResult.fields.serialNumber));
                
                response.putMap("fields", fields);
                
                WritableMap metadata = Arguments.createMap();
                metadata.putString("documentType", "ID_CARD");
                metadata.putInt("fieldCount", 5);
                metadata.putString("processingMethod", "ML_KIT");
                response.putMap("metadata", metadata);
                
                resolveOnUIThread(promise, response);
                
            } catch (Exception e) {
                Log.e(TAG, "Failed to retrieve OCR result", e);
                rejectOnUIThread(promise, "OCR_RETRIEVAL_ERROR", 
                        "Failed to retrieve result: " + e.getMessage(), null);
            }
        });
    }
    
    /**
     * Check if session token is valid
     * 
     * @param sessionToken Session token to validate
     * @param promise React Native promise
     */
    @ReactMethod
    public void isTokenValid(String sessionToken, Promise promise) {
        try {
            boolean isValid = secureStorage.isValid(sessionToken);
            resolveOnUIThread(promise, isValid);
        } catch (Exception e) {
            rejectOnUIThread(promise, "OCR_VALIDATION_ERROR", 
                    "Failed to validate token: " + e.getMessage(), null);
        }
    }
    
    /**
     * Cleanup expired tokens
     * 
     * @param promise React Native promise
     */
    @ReactMethod
    public void cleanupExpiredTokens(Promise promise) {
        executorService.submit(() -> {
            try {
                secureStorage.cleanupExpired();
                resolveOnUIThread(promise, true);
            } catch (Exception e) {
                rejectOnUIThread(promise, "OCR_CLEANUP_ERROR", 
                        "Failed to cleanup: " + e.getMessage(), null);
            }
        });
    }
    
    /**
     * Get current memory statistics
     * 
     * @param promise React Native promise
     */
    @ReactMethod
    public void getMemoryStats(Promise promise) {
        try {
            FrameProcessor.MemoryStats stats = frameProcessor.getMemoryStats();
            
            WritableMap response = Arguments.createMap();
            response.putDouble("maxMemoryMB", stats.maxMemoryMB);
            response.putDouble("totalMemoryMB", stats.totalMemoryMB);
            response.putDouble("usedMemoryMB", stats.usedMemoryMB);
            response.putDouble("freeMemoryMB", stats.freeMemoryMB);
            response.putDouble("usagePercent", stats.usagePercent);
            
            resolveOnUIThread(promise, response);
        } catch (Exception e) {
            rejectOnUIThread(promise, "OCR_MEMORY_ERROR", 
                    "Failed to get memory stats: " + e.getMessage(), null);
        }
    }
    
    // ==================== Private Helper Methods ====================
    
    /**
     * Perform OCR on bitmap (mock implementation)
     * Replace with actual OCR engine (ML Kit, Tesseract, etc.)
     */
    private OCRResult performOCR(Bitmap bitmap) throws Exception {
        // Simulate OCR processing
        Thread.sleep(100);
        
        return new OCRResult(
                "SUCCESS",
                0.93,
                new OCRFields(
                        new FieldData("12345678901", 0.95),
                        new FieldData("AHMET", 0.92),
                        new FieldData("YILMAZ", 0.94),
                        new FieldData("01.01.1990", 0.91),
                        new FieldData("A12B34567", 0.93)
                )
        );
    }
    
    /**
     * Create field data map for React Native
     */
    private WritableMap createFieldDataMap(FieldData fieldData) {
        WritableMap map = Arguments.createMap();
        map.putString("value", fieldData.value);
        map.putDouble("confidence", fieldData.confidence);
        return map;
    }
    
    /**
     * Create error map for React Native
     */
    private WritableMap createErrorMap(String code, String message, double confidence, boolean retryable) {
        WritableMap map = Arguments.createMap();
        map.putString("code", code);
        map.putString("message", message);
        map.putDouble("confidence", confidence);
        map.putBoolean("retryable", retryable);
        return map;
    }
    
    /**
     * Send event to React Native (no PII)
     */
    private void sendEvent(String eventName, WritableMap params) {
        getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }
    
    /**
     * Resolve promise on UI thread
     */
    private void resolveOnUIThread(Promise promise, Object value) {
        mainHandler.post(() -> promise.resolve(value));
    }
    
    /**
     * Reject promise on UI thread
     */
    private void rejectOnUIThread(Promise promise, String code, String message, WritableMap userInfo) {
        mainHandler.post(() -> {
            if (userInfo != null) {
                promise.reject(code, message, userInfo);
            } else {
                promise.reject(code, message);
            }
        });
    }
    
    /**
     * Cleanup resources
     */
    @Override
    public void onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy();
        frameProcessor.close();
        executorService.shutdown();
        Log.d(TAG, "OCRModule destroyed and resources cleaned up");
    }
    
    // ==================== Data Classes ====================
    
    public static class OCRResult {
        public final String status;
        public final double confidence;
        public final OCRFields fields;
        
        public OCRResult(String status, double confidence, OCRFields fields) {
            this.status = status;
            this.confidence = confidence;
            this.fields = fields;
        }
    }
    
    public static class OCRFields {
        public final FieldData tcNo;
        public final FieldData name;
        public final FieldData surname;
        public final FieldData birthDate;
        public final FieldData serialNumber;
        
        public OCRFields(FieldData tcNo, FieldData name, FieldData surname, 
                        FieldData birthDate, FieldData serialNumber) {
            this.tcNo = tcNo;
            this.name = name;
            this.surname = surname;
            this.birthDate = birthDate;
            this.serialNumber = serialNumber;
        }
    }
    
    public static class FieldData {
        public final String value;
        public final double confidence;
        
        public FieldData(String value, double confidence) {
            this.value = value;
            this.confidence = confidence;
        }
    }
}
